src/parser.rs

```diff
@@ -345,6 +345,15 @@ impl<'a> Parser<'a> {
             while self.consume_token(&Token::SemiColon) {
                 expecting_statement_delimiter = false;
             }
+            if self.peek_token()
```

Comment:
I don't understand the need for this code -- it seems like breaking parsing on `END` should only happen in the context of parsing a procedure.

A pattern used elsewhere in the code might be to conditionalize based on a parameter. Something like

```rust
    pub fn parse_statements(&mut self, in_procedure: bool) -> Result<Vec<Statement>, ParserError> {
```

For example